src/main/mondrian/olap/fun/MemberExtractingVisitor.java

```diff
@@ -0,0 +1,144 @@
+/*
+// This software is subject to the terms of the Eclipse Public License v1.0
+// Agreement, available at the following URL:
+// http://www.eclipse.org/legal/epl-v10.html.
+// You must accept the terms of that agreement to use this software.
+//
+// Copyright (c) 2002-2014 Pentaho Corporation..  All rights reserved.
+*/
+package mondrian.olap.fun;
+
+import mondrian.mdx.*;
+import mondrian.olap.*;
+import mondrian.olap.type.Type;
+
+import java.util.*;
+
+/**
+ * Visitor which collects any non-measure base members encountered while
+ * traversing an expression.
+ *
+ * This Visitor is used by the native set classes as well as the crossjoin
+ * optimizer (CrossjoinFunDef.nonEmptyList) to identify members within
+ * an expression which may conflict with members used as a constraint.
+ *
+ * If the boolean mapToAllMember is true, then any occurrences of
+ * a Dimension/Hierarchy/Level will result in the corresponding [All] member
+ * being added to the collection.  Likewise if a specific member is
+ * visited, the [All] member of it's corresponding hierarchy will be
+ * added.
+ *
+ * The mapToAllMember behavior will be used for any subexpression under
+ * one of the functions in the blacklist collection below.
+ */
+public class MemberExtractingVisitor extends MdxVisitorImpl {
+
+    private final Set<Member> memberSet;
+    private final ResolvedFunCallFinder finder;
+    private final Set<Member> activeMembers = new HashSet<Member>();
+    private final ResolvedFunCall call;
+    private final boolean mapToAllMember;
+
+    /**
+     * This list of functions are "blacklisted" because
+     * occurrence of a member/dim/level/hierarchy) within
+     * one of these expressions would result in a set of members
+     * that cannot be determined from the expression itself.
+     */
+    private static final String[] unsafeFuncNames = new String[] {
```

Comment:
I explored use of dependsOn(), but found many cases where it wasn't helpful to identify possible conflict.  For example, with a crossjoin of customers and products, if a measure on an opposing axis computes YTD sales, then [Time] is potentially conflicting, but will not be dependent.  Cases where the potentially conflicting member is directly in the crossjoin also may show no dependence.  For example if Gender.[F] is in the CJ it will be treated as a constant calc with dependence=false, even though it may be overridden elsewhere to be Gender.[M].

If you think there are other things worth trying please let me know.  Use of the unsafe list is pretty ugly but I couldn't come up with a better alternative.